<a href="https://colab.research.google.com/github/samateh10/football-dw-project/blob/main/Football_DW_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Football Player Performance — Data Warehouse Project

## Phase 2 — Data Quality Assessment and Cleaning

This notebook covers the full data pipeline for the football player
performance dataset. We assess data quality, clean the data, and
prepare it for loading into the reconciled database and data warehouse.

Dataset source: Kaggle — Football Player Stats across top European
and world leagues from 2016 to 2020.

In [28]:
# import the libraries we need
import pandas as pd
import numpy as np

In [29]:
# load the dataset into a dataframe
df = pd.read_csv('Data.csv')

In [30]:
# check the shape — how many rows and columns
print(df.shape)

(660, 15)


## Step 1 — Initial Data Inspection

Before assessing quality, we first look at the raw data to understand
its structure, column names, and data types.

In [31]:
# look at the first 5 rows of the dataset
df.head()

,Country,League,Club,Player Names,Matches_Played,Substitution,Mins,Goals,xG,xG Per Avg Match,Shots,OnTarget,Shots Per Avg Match,On Target Per Avg Match,Year
0,Spain,La Liga,(BET),Juanmi Callejon,19,16,1849,11,6.62,0.34,48,20,2.47,1.03,2016
1,Spain,La Liga,(BAR),Antoine Griezmann,36,0,3129,16,11.86,0.36,88,41,2.67,1.24,2016
2,Spain,La Liga,(ATL),Luis Suarez,34,1,2940,28,23.21,0.75,120,57,3.88,1.84,2016
3,Spain,La Liga,(CAR),Ruben Castro,32,3,2842,13,14.06,0.47,117,42,3.91,1.40,2016
4,Spain,La Liga,(VAL),Kevin Gameiro,21,10,1745,13,10.65,0.58,50,23,2.72,1.25,2016


In [32]:
# check all column names
print(df.columns.tolist())

['Country', 'League', 'Club', 'Player Names', 'Matches_Played', 'Substitution ', 'Mins', 'Goals', 'xG', 'xG Per Avg Match', 'Shots', 'OnTarget', 'Shots Per Avg Match', 'On Target Per Avg Match', 'Year']


In [33]:
# check the data type of each column
print(df.dtypes)

Country                     object
League                      object
Club                        object
Player Names                object
Matches_Played               int64
Substitution                 int64
Mins                         int64
Goals                        int64
xG                         float64
xG Per Avg Match           float64
Shots                        int64
OnTarget                     int64
Shots Per Avg Match        float64
On Target Per Avg Match    float64
Year                         int64
dtype: object


## Step 2 — Data Quality Assessment

We now formally assess the quality of the dataset across key dimensions
taught in the course — completeness, uniqueness, validity, and consistency.

In [34]:
# check for missing values in every column
print(df.isnull().sum())

Country                     0
League                      0
Club                       34
Player Names                0
Matches_Played              0
Substitution                0
Mins                        0
Goals                       0
xG                          0
xG Per Avg Match            0
Shots                       0
OnTarget                    0
Shots Per Avg Match         0
On Target Per Avg Match     0
Year                        0
dtype: int64


In [35]:
# check what percentage of each column is missing
print((df.isnull().sum() / len(df)) * 100)

Country                    0.000000
League                     0.000000
Club                       5.151515
Player Names               0.000000
Matches_Played             0.000000
Substitution               0.000000
Mins                       0.000000
Goals                      0.000000
xG                         0.000000
xG Per Avg Match           0.000000
Shots                      0.000000
OnTarget                   0.000000
Shots Per Avg Match        0.000000
On Target Per Avg Match    0.000000
Year                       0.000000
dtype: float64


In [36]:
# check for duplicate rows
print(df.duplicated().sum())

0


In [37]:
# check unique values in the Country column
print(df['Country'].unique())

['Spain' 'Italy' 'Germany' 'England' 'Brazil' 'France' 'USA' 'Portugal '
 ' Netherlands']


In [38]:
# check unique values in the League column
print(df['League'].unique())

['La Liga' 'Serie A' 'Bundesliga' 'Premier League'
 'Campeonato Brasileiro Série A' 'France Ligue 11' 'France Ligue 20'
 'France Ligue 2' 'France Ligue 12' 'France Ligue 9' 'France Ligue 15'
 'France Ligue 6' 'France Ligue 3' 'France Ligue 16' 'France Ligue 14'
 'France Ligue 4' 'France Ligue 1' 'France Ligue 10' 'France Ligue 7'
 'France Ligue 13' 'France Ligue 8' 'France Ligue 5' 'France Ligue 19'
 'France Ligue 18' 'France Ligue 17' 'MLS' 'Primeira Liga' 'Eredivisie']


## Step 3 — Data Cleaning

We fix all identified quality issues systematically before loading
data into the reconciled database.

In [39]:
# fix all column names — make them clean and consistent
df.columns = [
    'country', 'league', 'club', 'player_name',
    'matches_played', 'substitutions', 'minutes',
    'goals', 'xg', 'xg_per_avg_match', 'shots',
    'on_target', 'shots_per_avg_match',
    'on_target_per_avg_match', 'year'
]

In [40]:
# verify the new column names look correct
print(df.columns.tolist())

['country', 'league', 'club', 'player_name', 'matches_played', 'substitutions', 'minutes', 'goals', 'xg', 'xg_per_avg_match', 'shots', 'on_target', 'shots_per_avg_match', 'on_target_per_avg_match', 'year']


In [41]:
# fix leading and trailing spaces in country names
df['country'] = df['country'].str.strip()

In [42]:
# verify country names are now clean
print(df['country'].unique())

['Spain' 'Italy' 'Germany' 'England' 'Brazil' 'France' 'USA' 'Portugal'
 'Netherlands']


In [43]:
# fix the corrupted French league names
# all variations of "France Ligue..." should be "Ligue 1"
df['league'] = df['league'].apply(
    lambda x: 'Ligue 1' if str(x).startswith('France Ligue') else x
)

In [44]:
# verify league names are now clean
print(df['league'].unique())

['La Liga' 'Serie A' 'Bundesliga' 'Premier League'
 'Campeonato Brasileiro Série A' 'Ligue 1' 'MLS' 'Primeira Liga'
 'Eredivisie']


In [45]:
# fix the encoding error in the Brazilian league name
df['league'] = df['league'].str.replace(
    'Campeonato Brasileiro Série A',
    'Campeonato Brasileiro Série A'
)

In [46]:
# verify the Brazilian league name is now correct
print(df['league'].unique())

['La Liga' 'Serie A' 'Bundesliga' 'Premier League'
 'Campeonato Brasileiro Série A' 'Ligue 1' 'MLS' 'Primeira Liga'
 'Eredivisie']


In [47]:
# check the 34 missing club values — see what rows they are
print(df[df['club'].isnull()])

         country                         league club          player_name  \
10         Spain                        La Liga  NaN              Aduriz    
51       Germany                     Bundesliga  NaN          Mario Gomez   
82        Brazil  Campeonato Brasileiro Série A  NaN             Robinho    
83        Brazil  Campeonato Brasileiro Série A  NaN              Kempes    
85        Brazil  Campeonato Brasileiro Série A  NaN        Bruno Rangel    
89        Brazil  Campeonato Brasileiro Série A  NaN             Grafite    
92        Brazil  Campeonato Brasileiro Série A  NaN                Sassa   
96        Brazil  Campeonato Brasileiro Série A  NaN     Kleber Gladiador   
97        Brazil  Campeonato Brasileiro Série A  NaN             Pottker    
98        Brazil  Campeonato Brasileiro Série A  NaN      Jonathan Copete   
104        Spain                        La Liga  NaN         Rodrigo None   
149      Germany                     Bundesliga  NaN       Sandro Wagner    

### Handling Missing Club Values

34 rows have missing club codes representing 5.15% of the data.
Since all other columns for these rows are complete and valid,
we fill the missing club values with "Unknown" to preserve the
performance data rather than losing 34 valid player records.

In [48]:
# fill missing club values with Unknown
df['club'] = df['club'].fillna('Unknown')

In [49]:
# verify no missing values remain anywhere
print(df.isnull().sum())

country                    0
league                     0
club                       0
player_name                0
matches_played             0
substitutions              0
minutes                    0
goals                      0
xg                         0
xg_per_avg_match           0
shots                      0
on_target                  0
shots_per_avg_match        0
on_target_per_avg_match    0
year                       0
dtype: int64


In [50]:
# verify total rows is still 660 — no data was lost
print(len(df))

660


In [51]:
# final check — look at the clean dataset
df.head()

,country,league,club,player_name,matches_played,substitutions,minutes,goals,xg,xg_per_avg_match,shots,on_target,shots_per_avg_match,on_target_per_avg_match,year
0,Spain,La Liga,(BET),Juanmi Callejon,19,16,1849,11,6.62,0.34,48,20,2.47,1.03,2016
1,Spain,La Liga,(BAR),Antoine Griezmann,36,0,3129,16,11.86,0.36,88,41,2.67,1.24,2016
2,Spain,La Liga,(ATL),Luis Suarez,34,1,2940,28,23.21,0.75,120,57,3.88,1.84,2016
3,Spain,La Liga,(CAR),Ruben Castro,32,3,2842,13,14.06,0.47,117,42,3.91,1.40,2016
4,Spain,La Liga,(VAL),Kevin Gameiro,21,10,1745,13,10.65,0.58,50,23,2.72,1.25,2016


In [54]:
df.to_csv('clean_data.csv', index=False)

In [55]:
df.to_csv('clean_data.csv', index=False)
print('saved')

saved


## Step 4 — Populating the Reconciled Database

We connect to our MySQL database and create all the tables
for the reconciled database, then populate them with the
clean data following the correct order of foreign key
dependencies — Country first, then League, then Club,
then Player, and finally Performance.

In [52]:
# install the mysql connector library
!pip install mysql-connector-python

In [53]:
# import the connector
import mysql.connector

In [56]:
# connect to the online database
conn = mysql.connector.connect(
    host='sql7.freesqldatabase.com',
    user='sql7829613',
    password='MFGP1Izlsl',
    database='sql7829613',
    port=3306
)

# create cursor to run sql commands
cursor = conn.cursor()

print('Connected successfully')

Connected successfully


In [57]:
# create the Country table
cursor.execute('''
    CREATE TABLE IF NOT EXISTS Country (
        country_id INT AUTO_INCREMENT PRIMARY KEY,
        country_name VARCHAR(100) NOT NULL UNIQUE
    )
''')

print('Country table created')

Country table created


In [58]:
# create the League table
cursor.execute('''
    CREATE TABLE IF NOT EXISTS League (
        league_id INT AUTO_INCREMENT PRIMARY KEY,
        league_name VARCHAR(100) NOT NULL,
        country_id INT NOT NULL,
        FOREIGN KEY (country_id) REFERENCES Country(country_id)
    )
''')

print('League table created')

League table created


In [59]:
# create the Club table
cursor.execute('''
    CREATE TABLE IF NOT EXISTS Club (
        club_id INT AUTO_INCREMENT PRIMARY KEY,
        club_code VARCHAR(20),
        league_id INT NOT NULL,
        FOREIGN KEY (league_id) REFERENCES League(league_id)
    )
''')

print('Club table created')

Club table created


In [60]:
# create the Player table
cursor.execute('''
    CREATE TABLE IF NOT EXISTS Player (
        player_id INT AUTO_INCREMENT PRIMARY KEY,
        player_name VARCHAR(150) NOT NULL
    )
''')

print('Player table created')

Player table created


In [61]:
# create the Performance table
cursor.execute('''
    CREATE TABLE IF NOT EXISTS Performance (
        performance_id INT AUTO_INCREMENT PRIMARY KEY,
        player_id INT NOT NULL,
        club_id INT NOT NULL,
        year INT NOT NULL,
        matches_played INT,
        substitutions INT,
        minutes INT,
        goals INT,
        xg DECIMAL(6,2),
        xg_per_avg_match DECIMAL(6,2),
        shots INT,
        on_target INT,
        shots_per_avg_match DECIMAL(6,2),
        on_target_per_avg_match DECIMAL(6,2),
        FOREIGN KEY (player_id) REFERENCES Player(player_id),
        FOREIGN KEY (club_id) REFERENCES Club(club_id)
    )
''')

print('Performance table created')

Performance table created


In [62]:
# commit all table creations
conn.commit()
print('All tables created successfully')

All tables created successfully


In [63]:
# insert unique countries into Country table
countries = df['country'].unique()

for country in countries:
    cursor.execute(
        'INSERT IGNORE INTO Country (country_name) VALUES (%s)',
        (country,)
    )

conn.commit()
print('Countries inserted:', len(countries))

Countries inserted: 9


In [64]:
# insert unique leagues with their country foreign key
leagues = df[['league', 'country']].drop_duplicates()

for _, row in leagues.iterrows():
    cursor.execute(
        'SELECT country_id FROM Country WHERE country_name = %s',
        (row['country'],)
    )
    country_id = cursor.fetchone()[0]

    cursor.execute(
        'INSERT IGNORE INTO League (league_name, country_id) VALUES (%s, %s)',
        (row['league'], country_id)
    )

conn.commit()
print('Leagues inserted:', len(leagues))

Leagues inserted: 9


In [65]:
# insert unique clubs with their league foreign key
clubs = df[['club', 'league']].drop_duplicates()

for _, row in clubs.iterrows():
    cursor.execute(
        'SELECT league_id FROM League WHERE league_name = %s',
        (row['league'],)
    )
    league_id = cursor.fetchone()[0]

    cursor.execute(
        'INSERT IGNORE INTO Club (club_code, league_id) VALUES (%s, %s)',
        (row['club'], league_id)
    )

conn.commit()
print('Clubs inserted:', len(clubs))

Clubs inserted: 240


In [66]:
# insert unique players into Player table
players = df['player_name'].unique()

for player in players:
    cursor.execute(
        'INSERT IGNORE INTO Player (player_name) VALUES (%s)',
        (player,)
    )

conn.commit()
print('Players inserted:', len(players))

Players inserted: 444


In [68]:
# close the old connection and reconnect fresh
conn.close()

conn = mysql.connector.connect(
    host='sql7.freesqldatabase.com',
    user='sql7829613',
    password='MFGP1Izlsl',
    database='sql7829613',
    port=3306
)

# use buffered cursor to avoid unread result errors
cursor = conn.cursor(buffered=True)

print('Reconnected successfully')

Reconnected successfully


In [70]:
# find which clubs in the dataframe are missing from the Club table
for _, row in df.iterrows():
    cursor.execute(
        'SELECT club_id FROM Club WHERE club_code = %s',
        (row['club'],)
    )
    result = cursor.fetchone()
    if result is None:
        print('Missing club:', row['club'], '| League:', row['league'])
$0

Missing club: (chongqing dangdai lifan f.c) | League: Campeonato Brasileiro Série A


In [71]:
# check what clubs were inserted for the Brazilian league
cursor.execute(
    'SELECT club_id, club_code FROM Club WHERE league_id = (SELECT league_id FROM League WHERE league_name = %s)',
    ('Campeonato Brasileiro Série A',)
)

results = cursor.fetchall()
for r in results:
    print(r)

(70, '(GRE)')
(71, '(SAP)')
(72, 'Unknown')
(73, '(MNC)')
(74, '(CEA)')
(75, '(SAN)')
(76, '(HUR)')
(77, '(BOT)')
(78, '(FLA)')
(79, '(CAM)')
(80, '(FLU)')
(81, '(CFC)')
(100, '(OPE)')
(101, '(CRU)')
(102, '(BEN)')
(103, '(FOR)')
(104, '(COR)')
(105, '(SCR)')
(140, '(LYO)')
(141, '(ALA)')
(142, '(VDG)')
(143, '(RMA)')
(144, '(SJE)')
(145, '(TIG)')
(146, '(BAH)')
(147, 'Bruno')
(148, '(PAL)')
(200, '(INT)')
(201, '(RBB)')
(202, '(chongqing dangdai l')
(203, '(GOI)')
(231, '(CAP)')


In [72]:
# increase the club_code column size to handle long club names
cursor.execute(
    'ALTER TABLE Club MODIFY club_code VARCHAR(100)'
)

conn.commit()
print('Club table updated')

Club table updated


In [73]:
# update the truncated club code with the full name
cursor.execute(
    "UPDATE Club SET club_code = '(chongqing dangdai lifan f.c)' WHERE club_id = 202"
)

conn.commit()
print('Club code fixed')

Club code fixed


In [74]:
# verify the fix worked
cursor.execute(
    "SELECT club_id, club_code FROM Club WHERE club_id = 202"
)

print(cursor.fetchone())

(202, '(chongqing dangdai lifan f.c)')


In [75]:
# check if any other clubs are still missing
missing = []

for _, row in df.iterrows():
    cursor.execute(
        'SELECT club_id FROM Club WHERE club_code = %s',
        (row['club'],)
    )
    result = cursor.fetchone()
    if result is None:
        missing.append(row['club'])

print('Missing clubs:', set(missing))

Missing clubs: set()


In [76]:
# insert all performance records
for _, row in df.iterrows():
    cursor.execute(
        'SELECT player_id FROM Player WHERE player_name = %s',
        (row['player_name'],)
    )
    player_id = cursor.fetchone()[0]

    cursor.execute(
        'SELECT club_id FROM Club WHERE club_code = %s',
        (row['club'],)
    )
    club_id = cursor.fetchone()[0]

    cursor.execute('''
        INSERT INTO Performance (
            player_id, club_id, year, matches_played,
            substitutions, minutes, goals, xg,
            xg_per_avg_match, shots, on_target,
            shots_per_avg_match, on_target_per_avg_match
        ) VALUES (
            %s, %s, %s, %s, %s, %s, %s, %s,
            %s, %s, %s, %s, %s
        )
    ''', (
        player_id, club_id, row['year'],
        row['matches_played'], row['substitutions'],
        row['minutes'], row['goals'], row['xg'],
        row['xg_per_avg_match'], row['shots'],
        row['on_target'], row['shots_per_avg_match'],
        row['on_target_per_avg_match']
    ))

conn.commit()
print('Performance records inserted:', len(df))

Performance records inserted: 660


In [77]:
# verify row counts in all five tables
tables = ['Country', 'League', 'Club', 'Player', 'Performance']

for table in tables:
    cursor.execute(f'SELECT COUNT(*) FROM {table}')
    count = cursor.fetchone()[0]
    print(f'{table}: {count} rows')

Country: 9 rows
League: 9 rows
Club: 240 rows
Player: 444 rows
Performance: 1136 rows


In [78]:
# clear the Performance table
cursor.execute('DELETE FROM Performance')
conn.commit()
print('Performance table cleared')

Performance table cleared


In [79]:
# verify it is empty
cursor.execute('SELECT COUNT(*) FROM Performance')
print('Rows remaining:', cursor.fetchone()[0])

Rows remaining: 0


In [80]:
# reinsert all performance records cleanly
for _, row in df.iterrows():
    cursor.execute(
        'SELECT player_id FROM Player WHERE player_name = %s',
        (row['player_name'],)
    )
    player_id = cursor.fetchone()[0]

    cursor.execute(
        'SELECT club_id FROM Club WHERE club_code = %s',
        (row['club'],)
    )
    club_id = cursor.fetchone()[0]

    cursor.execute('''
        INSERT INTO Performance (
            player_id, club_id, year, matches_played,
            substitutions, minutes, goals, xg,
            xg_per_avg_match, shots, on_target,
            shots_per_avg_match, on_target_per_avg_match
        ) VALUES (
            %s, %s, %s, %s, %s, %s, %s, %s,
            %s, %s, %s, %s, %s
        )
    ''', (
        player_id, club_id, row['year'],
        row['matches_played'], row['substitutions'],
        row['minutes'], row['goals'], row['xg'],
        row['xg_per_avg_match'], row['shots'],
        row['on_target'], row['shots_per_avg_match'],
        row['on_target_per_avg_match']
    ))

conn.commit()
print('Performance records inserted:', len(df))

Performance records inserted: 660


In [81]:
# final verification of all five tables
tables = ['Country', 'League', 'Club', 'Player', 'Performance']

for table in tables:
    cursor.execute(f'SELECT COUNT(*) FROM {table}')
    count = cursor.fetchone()[0]
    print(f'{table}: {count} rows')

Country: 9 rows
League: 9 rows
Club: 240 rows
Player: 444 rows
Performance: 660 rows
